src/main/java/de/tum/in/www1/artemis/domain/ProgrammingExerciseTestCase.java

```diff
@@ -27,8 +28,9 @@
     @Column(name = "active")
     private Boolean active;
 
-    @Column(name = "after_due_date")
-    private Boolean afterDueDate;
+    @Enumerated(EnumType.STRING)
+    @Column(name = "visibility")
+    private TestCaseVisibility visibility;
```

Comment:
do we actually need the enum here as well?
Until now it was not necessary to determine whether feedback should be shown (in case of tests marked as after due date)

The issue is that it introduces a dependency to programming exercises.
We should try to have feedback as an exercise independent class.

Reply:
Here in the TestCase the visibility is definitely needed. I’m assuming this comment is about `Feedback.visibility`.

I have not used other exercise types or exams. Could a general visibility attribute for Feedbacks be useful for one of them in the future?
E.g. usable for an instructor to be able track how successful students are before the due date without having to reveal the Feedback (I don’t know if the approach where certain Feedbacks are just not saved in the database before the due date is used somewhere else other than ProgrammingExercises).

If not:
As far as I can tell, to link a Feedback back up to the corresponding TestCase at a later point one would have to go via Feedback -> Result -> Participation -> ProgrammingExercise -> ProgrammingExerciseTestCase and then check for `test_case.test_name = feedback.text` (automatic feedbacks) or something like `feedback.text not in test_case.test_name` (manual feedbacks).
In most places where Feedbacks are requested only the Participation is known. Therefore, for each request an additional database query fetching the TestCases for the Exercise would have to be made to then be able to filter Feedbacks accordingly.
In e.g. `ProgrammingExerciseParticipationResource.getParticipationWithLatestResultForStudentParticipation` all data is fetched from the database. However, the underlying query is already quite complex. A similar query that would also have to perform that join of Feedbacks to TestCase and already filter them based on requesting user, current time, and exercise due date to avoid a second database call would definitely make it far too complex, I think. Therefore, also for this an additional database call retrieving the TestCases for the Exercise and filtering Feedbacks in Java code would have to be performed.

If this trade-off of additional database calls is acceptable, I could remove `Feedback.visibility`.
The filtering functionality should then be bundled in a central class as it is used in multiple places. I’m not that familiar with Spring architecture. Should it be implemented as a service (`ProgrammingExerciseFeedbackFilterService`) or somehow else?